core/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/core/model/LocalUriInput.kt

```diff
@@ -0,0 +1,169 @@
+        const val MIN_PORT = 1
+        const val MAX_PORT = 65535
+        const val MIN_SOCKET_TIMEOUT = 1
+        const val MAX_SOCKET_TIMEOUT = 60
+
+        const val SCHEME_FIELD = "scheme"
+        const val HOST_FIELD = "host"
+        const val PORT_FIELD = "port"
+        const val PATH_FIELD = "path"
+        const val PARAMS_FIELD = "params"
+        const val URL_FIELD = "url"
+        const val CONNECTION_TIMEOUT_FIELD = "connection_timeout"
+        const val SOCKET_TIMEOUT_FIELD = "socket_timeout"
+        const val URI_FIELD = "uri"
+
+        val XCONTENT_REGISTRY = NamedXContentRegistry.Entry(Input::class.java, ParseField("uri"), CheckedFunction { parseInner(it) })
+
+        /**
+         * This parse function uses [XContentParser] to parse JSON input and store corresponding fields to create a [LocalUriInput] object
+         */
+        @JvmStatic @Throws(IOException::class)
+        private fun parseInner(xcp: XContentParser): LocalUriInput {
+            var scheme = "http"
+            var host = ""
+            var port: Int = -1
+            var path = ""
+            var params: Map<String, String> = mutableMapOf()
+            var url = ""
+            var connectionTimeout = MAX_CONNECTION_TIMEOUT
+            var socketTimeout = MAX_SOCKET_TIMEOUT
+
+            XContentParserUtils.ensureExpectedToken(XContentParser.Token.START_OBJECT, xcp.currentToken(), xcp)
+
+            while (xcp.nextToken() != XContentParser.Token.END_OBJECT) {
+                val fieldName = xcp.currentName()
+                xcp.nextToken()
+                when (fieldName) {
+                    SCHEME_FIELD -> scheme = xcp.text()
+                    HOST_FIELD -> host = xcp.text()
+                    PORT_FIELD -> port = xcp.intValue()
+                    PATH_FIELD -> path = xcp.text()
+                    PARAMS_FIELD -> params = xcp.mapStrings()
+                    URL_FIELD -> url = xcp.text()
+                    CONNECTION_TIMEOUT_FIELD -> connectionTimeout = xcp.intValue()
+                    SOCKET_TIMEOUT_FIELD -> socketTimeout = xcp.intValue()
+                }
+            }
+            return LocalUriInput(scheme, host, port, path, params, url, connectionTimeout, socketTimeout)
+        }
+    }
+
+    /**
+     * Constructs the [URI] either using [url] or using [scheme]+[host]+[port]+[path]+[query_params].
+     */
+    fun toConstructedUri(): URI {
+        return if (url.isEmpty()) {
+            val uriBuilder = URIBuilder()
+                .setScheme(scheme)
+                .setHost(host)
+                .setPort(port)
+                .setPath(path)
+            for (e in query_params.entries)
+                uriBuilder.addParameter(e.key, e.value)
+            uriBuilder.build()
+        } else {
+            URIBuilder(url).build()
+        }
+    }
+
+    /**
+     * Helper function to confirm at least [url], or [scheme]+[host]+[port]+[path]+[query_params] is defined.
```

Comment:
is it `either` or `alteast`? 

if its atleast then I'd think we need to check that 
either ``url`` and ``[scheme]+[host]+[port]+[path]+[query_params]`` is not empty and if they are both not empty then check they both are creating same URI?

If its either then do we need to check only one of them is non empty.


Looking at the code I am thinking either is what we are trying to do, but it seem to miss the validation for case that both are empty. We are only checking both are not not empty.

Reply:
Ah, thank you for pointing that out. Yes, we want **either** the `url` or the other fields to be provided, not both so we don't need to worry about a user providing two different URLs in one request. I've refactored this method to include that check.